### PR TITLE
Support cursor visibility based on input device type

### PR DESCRIPTION
**Pull request by freemangordon**
The logic is: if we receive motion event from a TS device, hide the cursor,
show it otherwise.

Most of the code was borrowed from Cordia h-d

Signed-off-by: Ivaylo Dimitrov <ivo.g.dimitrov.75@gmail.com>

### CODE REVIEW

File: src/main.c [hunk 1]
```diff
@@ -351,18 +362,120 @@ key_binding_func_key (MBWindowManager   *wm,
...
+
+          hd_xi_device *xi_dev = &g_array_index (xi_devices, hd_xi_device, id);
+
+          DeviceMotionNotify (dev, xi_motion_ev_type, ev_class);
+          g_array_append_val (eclass, ev_class);
+
+          xi_dev->is_ts = (info.type == atom_touchscreen);
+          xi_dev->dev = dev;
+#if 0
+          g_warning ("### %d/%u. %s(%d):%s.%d t:%d c:%d\n", i, eclass->len,
```
Comment (MerlijnWajer) [inline src/main.c:435]: Don't care much, but did you want to keep this here, in '#if 0' ?

**Review by MerlijnWajer — APPROVED** [review]

File: src/main.c (hunk 1 above)
Comment (freemangordon) [reply src/main.c:435]: It is very useful for debugging

File: src/main.c [hunk 2]
```diff
@@ -351,18 +362,120 @@ key_binding_func_key (MBWindowManager   *wm,
...
+      close_input_devices (dpy);
+      g_array_free (xi_devices, TRUE);
+      xi_devices = NULL;
+    }
+
+  if (atom_touchscreen == None)
+    atom_touchscreen = XInternAtom (dpy, XI_TOUCHSCREEN, True);
+
+  if (atom_touchscreen == None)
+    return;
```
Comment (spinal84) [inline src/main.c:403]: May be replaced with:
```
  if (atom_touchscreen == None) {
    atom_touchscreen = XInternAtom (dpy, XI_TOUCHSCREEN, True);

    if (atom_touchscreen == None)
      return;
  }
```

If we have `atom_touchscreen != None` (that is in all cases except the first call), we will just have two equal `if (...)` statements in a row, both of which will have FALSE.

File: src/main.c [hunk 3]
```diff
@@ -351,18 +362,120 @@ key_binding_func_key (MBWindowManager   *wm,
...
+  XDeviceInfo *devinfo;
+  int i, ndev;
+  GArray *eclass;
+
+  if (xi_devices)
+    {
+      close_input_devices (dpy);
+      g_array_free (xi_devices, TRUE);
+      xi_devices = NULL;
+    }
```
Comment (spinal84) [inline src/main.c:397]: I'd move this logic to `close_input_devices()` which may be renamed to `free_input_devices`.
So we just call `free_input_devices()` here and on hildon-desktop termination.

File: src/main.c [hunk 4]
```diff
@@ -351,18 +362,120 @@ key_binding_func_key (MBWindowManager   *wm,
...
+  int i;
+
+  for (i = 0; i < xi_devices->len; i++)
+    {
+      hd_xi_device *xi_dev = &g_array_index (xi_devices, hd_xi_device, i);
+
+      if (xi_dev->dev)
+        {
+          XCloseDevice (dpy, xi_dev->dev);
+          xi_dev->dev = NULL;
```
Comment (spinal84) [inline src/main.c:378]: Why is this needed?

File: src/main.c [hunk 5]
```diff
@@ -351,18 +362,120 @@ key_binding_func_key (MBWindowManager   *wm,
...
+  devinfo = XListInputDevices (dpy, &ndev);
+
+  eclass = g_array_new (FALSE, FALSE, sizeof (XEventClass));
+  xi_devices = g_array_new (FALSE, TRUE, sizeof (hd_xi_device));
+
+  for (i = 0; i < ndev; i++)
+    {
+      XDeviceInfo info = devinfo[i];
+
+      if (info.use == IsXExtensionPointer)
```
Comment (spinal84) [inline src/main.c:418]: Shouldn't it be
`if (info.use == IsXPointer || info.use == IsXExtensionPointer)`?

File: src/main.c [hunk 6]
```diff
@@ -351,18 +362,120 @@ key_binding_func_key (MBWindowManager   *wm,
   hd_dbus_send_event (s);
 }
 
+/* Close all input devices previously opened */
+static void
+close_input_devices (Display *dpy)
```
Comment (spinal84) [inline src/main.c:367]: The call to this function is missing on hildon-desktop termination.

File: src/main.c (hunk 3 above)
Comment (freemangordon) [reply src/main.c:397]: Yes, it can be done as you say, but it is a matter of preference, There is no gain.

Also, freeing function a() allocated memory in function b() is bad coding practice, that's why it is done like that.

File: src/main.c (hunk 2 above)
Comment (freemangordon) [reply src/main.c:403]: I don;t really like such nested ifs, that's why I do it like that. Regarding the performance - I can bet compiler will optimize.

File: src/main.c (hunk 4 above)
Comment (freemangordon) [reply src/main.c:378]: because it will prevent SEGFAULTs if somebody decides to (wrongly) call the function for the second time. Also, we better have no dangling pointers around.

File: src/main.c (hunk 5 above)
Comment (freemangordon) [reply src/main.c:418]: I think IsXPointer is for core pointer devices, i.e. those that are not a real mice or TSes. Do you think IsXPointer shall be included and why?

File: src/main.c (hunk 6 above)
Comment (freemangordon) [reply src/main.c:367]: Yes, but there is no graceful termination of hildon-desktop I was able to find.

File: src/main.c (hunk 5 above)
Comment (spinal84) [reply src/main.c:418]: Isn't it possible that the user configure a touchscreen as a core pointer device?

Comment (spinal84) [reply src/main.c:418]: > i.e. those that are not a real mice or TSes.

Could you please share a link to documentation about this?
I wasn't aware of it (core pointer devices to be virtual) and couldn't fine an information.

Comment (spinal84) [reply src/main.c:418]: From xorg.conf (5) manpage:

`If hotplugging is disabled, there will normally be at least two: one for the core (primary) keyboard and one for the core pointer. If either of these two is missing, a default configuration for the missing ones will be used. In the absence of an explicitly specified core input device, the first InputDevice marked as CorePointer (or CoreKeyboard) is used.`

File: src/main.c (hunk 6 above)
Comment (spinal84) [reply src/main.c:367]: What about the code block coming after the call to `gtk_main()` in the main function of main.c file?

File: src/main.c (hunk 2 above)
Comment (spinal84) [reply src/main.c:403]: > Regarding the performance - I can bet compiler will optimize.

According to my tests with `gcc -O2 -S`, it generates additional assembler instruction for the case without nested if's (on amd64 arch). When gcc doesn't use -O2, the instruction set size is the same, but it executes the check inside the if statement for two times, just like it's done in the code.

Anyway, if you don't like nested 'if' so much, let's leave it intact.

File: src/main.c (hunk 5 above)
Comment (freemangordon) [reply src/main.c:418]: The point is that we wan't to make a difference whether a device is mouse or ts. Core pointers' type atom is null, so it makes no sense checking them. So, if we have core pointers only, cursor will be hidden initially and that's it.

File: src/main.c (hunk 6 above)
Comment (freemangordon) [reply src/main.c:367]: ok, I'll put a call to close_input_devices() there, but I doubt it will ever be reached.
